app/src/main/java/com/worldsnas/starplayer/ExoPlayerService.kt

```diff
@@ -0,0 +1,387 @@
+import android.text.TextUtils
+import androidx.core.app.NotificationCompat
+import com.google.android.exoplayer2.*
+import com.google.android.exoplayer2.source.*
+import com.google.android.exoplayer2.trackselection.TrackSelectionArray
+import com.google.android.exoplayer2.upstream.DefaultHttpDataSourceFactory
+import com.worldsnas.starplayer.model.Music
+import java.util.*
+
+class ExoPlayerService : Service(), Player.EventListener {
+
+
+    lateinit var mediaSource: ConcatenatingMediaSource
+
+
+    lateinit var player: SimpleExoPlayer
+
+    private var playbackPosition: Long = 0
+    private lateinit var musicList: ArrayList<Music>
+    private var currentMusicPosition: Int = 0
+
+
+    override fun onDestroy() {
+        super.onDestroy()
+        releasePlayer()
+        releasePlayer()
+    }
+
+
+    private fun play(trackModel: Music) {
+        musicList = ArrayList<Music>()
+        musicList.add(trackModel)
+        play(musicList)
+    }
+
+    private fun play(trackModels: ArrayList<Music>) {
+        playbackPosition = 0
+        if (trackModels.size > 0) {
+            val mediaSources =
+                arrayOfNulls<MediaSource>(trackModels.size)
+            for (i in trackModels.indices) {
+                mediaSources[i] =
+                    buildMediaSource(Uri.parse(trackModels[i].address))
+            }
+            mediaSource = ConcatenatingMediaSource(*mediaSources)
+            player.prepare(mediaSource)
+            startPlayer()
+        }
+    }
+
+
+    private fun pausePlayer() {
+        playbackPosition = player.currentPosition
+        currentMusicPosition = player.currentWindowIndex
+        player.playWhenReady = false
+        showCurrentTrackNotification(musicList[player.currentWindowIndex])
+    }
+
+    private fun startPlayer() {
+        player.seekTo(currentMusicPosition, playbackPosition)
+        player.playWhenReady = true
+        showCurrentTrackNotification(musicList[player.currentWindowIndex])
+    }
+
+    private fun initializePlayer() {
+        player = SimpleExoPlayer.Builder(this).build()
+        player.setAudioAttributes(
+            com.google.android.exoplayer2.audio.AudioAttributes.Builder()
+                .setUsage(C.USAGE_MEDIA)
+                .setContentType(C.CONTENT_TYPE_SPEECH)
+                .build(), true
+        )
+
+
+        player.addListener(this)
+
+    }
+
+    private fun buildMediaSource(uri: Uri): MediaSource {
+        val userAgent = "star_player"
+        return ProgressiveMediaSource.Factory(DefaultHttpDataSourceFactory(userAgent))
+            .createMediaSource(uri)
+    }
+
+    override fun onCreate() {
+        super.onCreate()
+        initializePlayer()
+    }
+
+    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
+
+        when (intent.action) {
+            PLAY_ACTION -> {
+
+                startPlayer()
+            }
+            PAUSE_ACTION -> pausePlayer()
+            START_TRACKS_ARRAY_ACTION -> {
```

Comment:
in ba list esh hich farghi nadare benazaram to factory function ha handle konim behtar bashe nist?

Reply:
manzureto motevageh nashodam